Changes for Dedicated Host with Landing Zone VSI - Reg

Louies-Jhony opened the pull request:
### Description

<!--- Replace this text with a summary of the changes in this PR. Include why the changes are needed and context about the changes. List required dependencies. If there is a Git issue for the change, please link to it. --->

### Release required?
<!--- Identify the type of release. For information about the changes in a semantic versioning release, see [Release versioning](https://terraform-ibm-modules.github.io/documentation/#/versioning). --->

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content

<!--- If a release is required, replace this text with information that users need to know about the release. Write the release notes to help users understand the changes, and include information about how to update from the previous version.

Your notes help the merger write the commit message for the PR that is published in the release notes for the module. --->

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

Louies-Jhony commented:
/run pipeline

Louies-Jhony commented:
/run pipeline

Louies-Jhony commented:
/run pipeline

Louies-Jhony commented:
/run pipeline

Louies-Jhony commented:
/run pipeline

Louies-Jhony commented:
/run pipeline

Louies-Jhony commented:
/run pipeline

Louies-Jhony commented:
/run pipeline

Louies-Jhony commented:
/run pipeline

Louies-Jhony commented:
/run pipeline

Louies-Jhony commented:
/run pipeline

ocofaigh commented on variables.tf:
```diff
@@ -502,3 +502,26 @@ variable "use_legacy_network_interface" {
+  default     = null
+  description = "ID of the dedicated host for hosting the VSI's"
+
+  validation {
```
In order to use this validation, users need to be on terraform version >=1.9.0. So can we bump the required terraform version in the module's version.tf as all submodules and examples?

Louies-Jhony replied:
Updated to 1.9.0

ocofaigh commented on examples/complete/variables.tf:
```diff
@@ -51,3 +51,9 @@ variable "secondary_use_vsi_security_group" {
   default     = false
 }
+
+variable "enable_dedicated_host" {
```
no need to expose a new variable in the example - we don't want to make examples flexible solutions. Suggest to just hard code this to true in the example main.tf

ocofaigh replied:
Can you also update the `examples/complete/README.md` file to mention about dedicated host

Louies-Jhony replied:
Updated

ocofaigh commented on variables.tf:
```diff
@@ -502,3 +502,26 @@ variable "use_legacy_network_interface" {
+variable "enable_dedicated_host" {
+  type        = bool
+  default     = false
+  description = "Setting this option to true will enable dedicated hosts for the VSI's, the default value is set to false. Refer [Understanding Dedicated Hosts](https://cloud.ibm.com/docs/vpc?topic=vpc-creating-dedicated-hosts-instances&interface=ui#about-dedicated-hosts) for more details"
```
update the description to mention that when set to true, a value is required for the `dedicated_host_id` input

Louies-Jhony replied:
Updated

ocofaigh commented on variables.tf:
```diff
@@ -502,3 +502,26 @@ variable "use_legacy_network_interface" {
+
+  validation {
+    condition     = var.enable_dedicated_host == false || (var.enable_dedicated_host == true && var.dedicated_host_id != null)
+    error_message = "When enable_dedicated_host is set to true, provide dedicated_host_id or new dedicated host will be created."
```
new dedicated host will be created? Should this say "new dedicated host will **not** be created"

Louies-Jhony replied:
Updated

ocofaigh commented on variables.tf:
```diff
@@ -502,3 +502,26 @@ variable "use_legacy_network_interface" {
+variable "dedicated_host_id" {
+  type        = string
+  default     = null
+  description = "ID of the dedicated host for hosting the VSI's"
```
Update the description to mention that `enable_dedicated_host` needs to be set to true if passing a dedicated host ID

Louies-Jhony replied:
Updated

ocofaigh commented on examples/complete/main.tf:
```diff
@@ -226,3 +239,65 @@ module "slz_vsi" {
+}
+
+#############################################################################
+# Dedicated Host
```
maybe put the code for dedicated host above the VSI code since host is needed before VSI

ocofaigh commented on examples/complete/main.tf:
```diff
@@ -24,7 +24,7 @@ module "resource_group" {
 module "key_protect_all_inclusive" {
   source                    = "terraform-ibm-modules/kms-all-inclusive/ibm"
-  version                   = "4.19.2"
+  version                   = "4.19.1"
```
this is an older version to whats in main branch - suggest to change it back

Louies-Jhony replied:
Updated

ocofaigh reviewed (changes requested):
Left some comments. Can you also expose this feature in the fscloud submodule code (found in `modules/fscloud`)

Louies-Jhony commented:
/run pipeline

Louies-Jhony commented:
/run pipeline

Louies-Jhony commented:
/run pipeline

Louies-Jhony commented:
@ocofaigh 

For dedicated host testing,  the existing VSI module create VSI's with placement group across zones. Since VSI can be place  either in a placement group or a dedicated host, I have introduced a separate module for VSI which will create a VSI  in a dedicated host on zone-1.

On the PR test go file, for complete and upgrade test case, we are using the same setup.(i.e same folder COMPLETE)
All test functions runs in parallel which created quota issue. Hence dedicated host can be included on any of the test case.

I have addressed all other comments.

Please advice.

Louies-Jhony commented:
/run pipeline

Aashiq-J commented:
This PR no longer valid.
Changes made in https://github.com/terraform-ibm-modules/terraform-ibm-landing-zone-vsi/pull/780
